src/cargo/ops/cargo_compile.rs

```diff
@@ -568,15 +569,34 @@ fn generate_targets<'a>(
 
     // Helper for creating a Unit struct.
     let new_unit =
-        |pkg: &'a Package, target: &'a Target, mode: CompileMode, profile_for: ProfileFor| {
-            let mode = match mode {
+        |pkg: &'a Package, target: &'a Target, target_mode: CompileMode| {
+            let profile_for = if mode == CompileMode::Test {
+                // NOTE: The ProfileFor here is subtle.  If you have a profile
+                // with `panic` set, the `panic` flag is cleared for tests and
+                // their dependencies.  If we left this as an "Any" profile,
+                // then the lib would get compiled three times (once with
+                // panic, once without, and once with --test).
+                //
+                // This would cause a problem for Doc tests, which would fail
+                // because `rustdoc` would attempt to link with both libraries
+                // at the same time. Also, it's probably not important (or
+                // even desirable?) for rustdoc to link with a lib with
+                // `panic` set.
+                //
+                // As a consequence, Examples and Binaries get compiled
+                // without `panic` set.  This probably isn't a bad deal.
```

Comment:
Ah yeah, this is exactly what I was thinking about... Hm, compiling binaries with different panic setting actually seems like a big deal to me...

Reply:
Yea, it's a tricky issue, I was going to ask about it.  The immediate problem is that doc tests attempt to link with everything in `Context.libraries`.  I'm uncertain what the easiest way to fix that would be.  One thought I had would be to change `run_doc_tests` to only grab the dependencies for the `CompileMode::Doctest` unit and remove `Context.libraries`.  There would be some other minor things to accommodate that, but I think it should work.  I can do it now or later, let me know.  It would remove this weird special-case.

Reply:
Would it be difficult to preserve exactly the behavior (and bugs :) ) we have today? I think this would be the best way forward for now, because changes which both add unstable features and fix/modify stable behavior are tricky...

Reply:
Yea, it should be preserving the old behavior.  I have a few changes I haven't pushed, yet, that ensure that.  I'm keeping a list of behavior changes, and I'll post it soon.  A few are clearly bug fixes, but a couple are questionable.  We can review them when I post it.